File: 2024/javascript/day01/main.js
import { readFileSync } from 'node:fs';

// Read values from file
const lines = readFileSync('./input.txt')
  .toString()
  .split('\n');
const left = /** @type {number[]} */ ([]);
const right = /** @type {number[]} */([]);
const pattern = /(\d+)\s+(\d+)/;
for(const line of lines) {
  const match = line.split(pattern)
  if (!match) {
    continue;
  }

  left.push(Number(match[1]));
  right.push(Number(match[2]));
}

// sort lists
const leftSorted = left.toSorted((a, b) => a - b);
const rightSorted = right.toSorted((a, b) => a - b);

// distance
const distance =
  leftSorted
    .map((_, index) => Math.abs(leftSorted[index] - rightSorted[index]))
    .reduce((a, b) => a + b)
console.log(`Solution 1: Distance between lists: ${distance}`);

// score
let score = 0;
for (const value of leftSorted) {
  const total = rightSorted.filter((candidate) => candidate === value).length;
  score += value * total;
}
console.log(`Solution 2: Similarity score between lists: ${score}`);

File: 2024/kotlin/src/day01/Main.kt
package day01

import java.io.File
import kotlin.math.abs

fun readValues(): Pair<List<Int>, List<Int>> {
  val pattern = """(\d+)\s+(\d+)""".toRegex()
  val left = mutableListOf<Int>()
  val right = mutableListOf<Int>()
  for (line in File("src/day01/input.txt").readLines()) {
    val match = pattern.find(line)
    val values = match?.groupValues ?: listOf()
    if (values.isEmpty()) {
      continue
    }
    
    left.add(values[1].toInt())
    right.add(values[2].toInt())
  }
  
  return Pair(left.sorted(), right.sorted())
}

fun main() {
  val (left, right) = readValues()
  
  // Calculate distances
  val distances = mutableListOf<Int>()
  for (i in left.indices) {
    val distance = abs(left[i] - right[i])
    distances.add(distance)
  }
  
  val totalDistances = distances.reduce { a, b -> a + b }
  println("Total distances between lists = $totalDistances")

  // Calculate similarity score
  var totalSimilarityScore = 0
  for (value in left) {
    val total = right.filter { it == value }.sum()
    val score = value * total
    totalSimilarityScore += score
  }
  
  println("Total similarity score between lists = $totalSimilarityScore")
}
